distributed/tests/test_preload.py

```diff
@@ -161,61 +155,40 @@ async def test_preload_import_time(cleanup):
         del backends["foo"]
 
 
-class MyHandler(web.RequestHandler):
-    def get(self):
-        self.write(
-            """
-def dask_setup(dask_server):
-    dask_server.foo = 1
-""".strip()
+@gen_test()
+async def test_web_preload():
+    def preload():
+        with server_sock.accept()[0] as client_sock, client_sock.makefile("rwb") as f:
+            assert f.readline() == b"GET /preload HTTP/1.1\r\n"
+            f.write(
+                b"HTTP/1.1 200 OK"
+                b"\r\nContent-Length: 53"
+                b"\r\n\r\n"
+                b"def dask_setup(dask_server):"
+                b"\n    dask_server.foo = 1"
+                b"\n"
+            )
+
+    async def test():
+        port = server_sock.getsockname()[1]
+        with captured_logger("distributed.preloading") as log:
+            async with Scheduler(
+                host="localhost",
+                preload=[f"http://127.0.0.1:{port}/preload"],
+            ) as s:
+                assert s.foo == 1
+        assert (
+            re.match(
+                r"(?s).*Downloading preload at http://127.0.0.1:\d+/preload\n"
+                r".*Run preload setup function: http://127.0.0.1:\d+/preload\n"
+                r".*",
+                log.getvalue(),
+            )
+            is not None
         )
 
-
-def create_preload_application():
-    app = web.Application([(r"/preload", MyHandler)])
-    server = app.listen(12345, address="127.0.0.1")
-    tornado.ioloop.IOLoop.instance().start()
-
-
-@pytest.fixture
-def scheduler_preload():
-    p = multiprocessing.Process(target=create_preload_application)
-    p.start()
-    start = time()
-    while not p.is_alive():
-        if time() > start + 5:
-            raise AssertionError("Process didn't come up")
-        sleep(0.5)
-    # Make sure we can query the server
-    start = time()
-    request = urllib.request.Request("http://127.0.0.1:12345/preload", method="GET")
-    while True:
-        try:
-            response = urllib.request.urlopen(request)
-            if response.status == 200:
-                break
-        except urllib.error.URLError as e:
-            if time() > start + 10:
-                raise AssertionError("Webserver didn't come up", e)
-            sleep(0.5)
-
-    yield
-    p.kill()
-    p.join(timeout=5)
-
-
-@pytest.mark.skipif(
-    MACOS and PY_VERSION == (3, 7), reason="HTTP Server doesn't come up"
-)
-@pytest.mark.asyncio
-async def test_web_preload(cleanup, scheduler_preload):
-    with captured_logger("distributed.preloading") as log:
-        async with Scheduler(
-            host="localhost",
-            preload=["http://127.0.0.1:12345/preload"],
-        ) as s:
-            assert s.foo == 1
-    assert "12345/preload" in log.getvalue()
+    with socket.create_server(("127.0.0.1", 0)) as server_sock:
+        await asyncio.gather(to_thread(preload), test())
```

Comment:
This seems to be a  more robust expression of the previous test, achieved by:

1. Avoiding creating the preload application in a separate process
2. Allowing a dynamically allocated port